crates/proto/src/rr/rdata/opt.rs

```diff
@@ -219,12 +219,21 @@ pub fn read(decoder: &mut BinDecoder, rdata_length: Restrict<u16>) -> ProtoResul
                     .map(|u| u as usize)
                     .verify_unwrap(|u| *u <= rdata_length)
                     .map_err(|_| ProtoError::from("OPT value length exceeds rdata length"))?;
-                state = OptReadState::Data {
-                    code,
-                    length,
-                    // TODO: this cean be replaced with decoder.read_vec(), right?
-                    //  the current version allows for malformed opt to be skipped...
-                    collected: Vec::<u8>::with_capacity(length),
+                // If we know that the length is 0, we can avoid the `OptReadState::Data` state
+                // and directly add the option to the map.
+                // The data state does not process 0-length correctly, since it always reads at
+                // least 1 byte, thus making the length check fail.
+                state = if length == 0 {
```

Comment:
Could it be simple enough to just change the Data state to only push and pop in the else statement?

The other option is to change it from pop() to read_vec(), I don't think it needs to loop over pop...

Reply:
That would be much nicer, however the `OptReadState::Data` will not always be reached. If the EDNS record ends in an empty (0-length) option, then the loop will exit before processing it. That is, because `decoder.index() - start_idx` has the same value as `rdata_length`, since alreay all bytes of the rdata are read. It would require changing the while-loop condition where I am unsure how to do that correctly.

Reply:
Could change it to >= and then handle the = case in the various states? Not sure how unholy that would be.